packages/theia-integration/src/browser/diagram/theia-marker-manager.ts

```diff
@@ -40,20 +42,52 @@ export function connectTheiaMarkerManager(
     }
 }
 
+type ValidationMarker = Marker & { reason?: string };
+
 class DiagnosticMarkers {
-    protected diagnotic2marker = new Map<Diagnostic, Marker>();
+    protected diagnotic2marker = new Map<Diagnostic, ValidationMarker>();
     get size(): number {
         return this.diagnotic2marker.size;
     }
-    all(): IterableIterator<Marker> {
+    all(): IterableIterator<ValidationMarker> {
         return this.diagnotic2marker.values();
     }
-    marker(diagnostic: Diagnostic): Marker | undefined {
+    marker(diagnostic: Diagnostic): ValidationMarker | undefined {
         return this.diagnotic2marker.get(diagnostic);
     }
-    add(diagnostic: Diagnostic, marker: Marker): Map<Diagnostic, Marker> {
+    add(diagnostic: Diagnostic, marker: ValidationMarker): Map<Diagnostic, ValidationMarker> {
         return this.diagnotic2marker.set(diagnostic, marker);
     }
+    getMarkerByOrigin(origin?: string): ValidationMarker[] {
+        const markers = [];
+        for (const diagnostic of this.diagnotic2marker.keys()) {
+            const marker = this.diagnotic2marker.get(diagnostic);
+            if (marker && marker.reason === origin) {
+                markers.push(marker);
+            }
+        }
+        return markers;
+    }
+    getByOrigin(origin?: string): Diagnostic[] {
+        const diagnostics = [];
+        for (const diagnostic of this.diagnotic2marker.keys()) {
```

Comment:
Nitpick:
You could use `forEach` here instead of iterating over the keys. This way you don't have to retrieve the value from the map and can avoid the `undefined` check:

```ts
    this.diagnotic2marker.forEach((marker,diagnostic)=>{
            if (marker.reason === origin){
                diagnostics.push(diagnostic);
            }
        });
```